### ISSUE TITLE
Check for byte string when using SNI

### ISSUE THREAD

**Issue by tsani** [pyca/pyopenssl#515]
I'm (indirectly) using pyOpenSSL version 16.0.0. When using SNI, there's a type check that the name to set is in fact a byte string. I'm also using requests (and consequently urllib3); although I'm passing in byte strings to `requests.get`, at some point the byte string becomes changed to a unicode string, so the server name fails the type check.

I'm as of yet unable to determine at what point in the call stack my byte string is coerced to a unicode string. Perhaps as a workaround for the very common case of a server name consisting only of ASCII anyway, pyOpenSSL can try to coerce the unicode string to ASCII before bailing.

Here's the stack trace of the error I get specifically. I've omitted the first few frames, which come from my code.

```
  File "/home/oohlala/oohlala_env/lib/python2.6/site-packages/requests/api.py", line 69, in get
    return request('get', url, params=params, **kwargs)
  File "/home/oohlala/oohlala_env/lib/python2.6/site-packages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/home/oohlala/oohlala_env/lib/python2.6/site-packages/requests/sessions.py", line 465, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/oohlala/oohlala_env/lib/python2.6/site-packages/requests/sessions.py", line 573, in send
    r = adapter.send(request, **kwargs)
  File "/home/oohlala/oohlala_env/lib/python2.6/site-packages/requests/adapters.py", line 370, in send
    timeout=timeout
  File "/home/oohlala/oohlala_env/lib/python2.6/site-packages/requests/packages/urllib3/connectionpool.py", line 544, in urlopen
    body=body, headers=headers)
  File "/home/oohlala/oohlala_env/lib/python2.6/site-packages/requests/packages/urllib3/connectionpool.py", line 341, in _make_request
    self._validate_conn(conn)
  File "/home/oohlala/oohlala_env/lib/python2.6/site-packages/requests/packages/urllib3/connectionpool.py", line 761, in _validate_conn
    conn.connect()
  File "/home/oohlala/oohlala_env/lib/python2.6/site-packages/requests/packages/urllib3/connection.py", line 238, in connect
    ssl_version=resolved_ssl_version)
  File "/home/oohlala/oohlala_env/lib/python2.6/site-packages/requests/packages/urllib3/contrib/pyopenssl.py", line 279, in ssl_wrap_socket
    cnx.set_tlsext_host_name(server_hostname)
  File "/home/oohlala/oohlala_env/lib/python2.6/site-packages/OpenSSL/SSL.py", line 1234, in set_tlsext_host_name
    raise TypeError("name must be a byte string")
```

**Comment by mithrandi** [comment]
We debugged this in #cryptography-dev and it looks like this was an issue in urllib3 that is already fixed; upgrading to requests 2.10.0 or higher fixes the issue (and requests 2.10.0 bundled a newer urllib3).

**Comment by Lukasa** [comment]
Yup, this was definitionally a urllib3 bug: PyOpenSSL sets what its API will and won't accept, and urllib3 has to conform. I believe that urllib3+requests does conform in 2.11, but please let me know if that turns out not to be true.
